torch/csrc/distributed/c10d/Functional.cpp

```diff
@@ -0,0 +1,258 @@
+
+c10d::ReduceOp to_reduce_op(const std::string& reduce_op) {
+  auto it = str_to_reduce_op.find(reduce_op);
+  TORCH_CHECK(
+      it != str_to_reduce_op.end(), "Unrecognized reduce_op: ", reduce_op);
+  return it->second;
+}
+
+at::Tensor all_reduce_(
+    at::Tensor input,
+    const std::string& reduce_op,
+    const std::string& group_name) {
+  c10d::AllreduceOptions opts;
+  opts.reduceOp = to_reduce_op(reduce_op);
+
+  std::vector<at::Tensor> inputs{input};
+  auto group = c10d::resolve_process_group(group_name);
+  auto work = group->allreduce(inputs, opts);
+  c10d::RankLocal<WorkRegistry>::get().register_work(input, work);
+  return input;
+}
+
+at::Tensor all_reduce(
+    const at::Tensor& input,
+    const std::string& reduce_op,
+    const std::string& group_name) {
+  auto output = input.clone();
+  return all_reduce_(output, reduce_op, group_name);
+}
+
+std::vector<at::Tensor> all_reduce_coalesced_(
+    std::vector<at::Tensor> inputs,
+    const std::string& reduce_op,
+    const std::string& group_name) {
+  c10d::AllreduceCoalescedOptions opts;
+  opts.reduceOp = to_reduce_op(reduce_op);
+
+  auto group = c10d::resolve_process_group(group_name);
+  auto work = group->allreduce_coalesced(inputs, opts);
+  for (const auto& tensor : inputs) {
+    c10d::RankLocal<WorkRegistry>::get().register_work(tensor, work);
+  }
+  return inputs;
+}
+
+std::vector<at::Tensor> all_reduce_coalesced(
+    const std::vector<at::Tensor>& inputs,
+    const std::string& reduce_op,
+    const std::string& group_name) {
+  std::vector<at::Tensor> outputs;
+  for (const auto& tensor : inputs) {
+    outputs.push_back(tensor.clone());
+  }
+  return all_reduce_coalesced_(outputs, reduce_op, group_name);
+}
+
+at::Tensor allocate_all_gather_output(
+    const at::Tensor& input,
+    int64_t group_size) {
+  auto output_size = input.sizes().vec();
+  output_size[0] *= group_size;
+  return at::empty(
+      output_size,
+      at::TensorOptions().dtype(input.dtype()).device(input.device()));
+}
+
+std::vector<at::Tensor> all_gather_into_tensor_coalesced(
+    const std::vector<at::Tensor>& inputs,
+    const int64_t group_size,
+    const std::string& group_name) {
+  std::vector<at::Tensor> outputs;
+  for (const auto& tensor : inputs) {
+    outputs.push_back(allocate_all_gather_output(tensor, group_size));
+  }
+
+  auto group = c10d::resolve_process_group(group_name);
+  auto work = group->allgather_into_tensor_coalesced(
+      outputs, const_cast<std::vector<at::Tensor>&>(inputs));
+  for (const auto& tensor : outputs) {
+    c10d::RankLocal<WorkRegistry>::get().register_work(tensor, work);
+  }
+  return outputs;
+}
+
+at::Tensor all_gather_into_tensor(
+    const at::Tensor& input,
+    const int64_t group_size,
+    const std::string& group_name) {
+  std::vector<at::Tensor> inputs{input};
+  return all_gather_into_tensor_coalesced(inputs, group_size, group_name)[0];
+}
+
+at::Tensor allocate_reduce_scatter_output(
```

Comment:
nit: maybe make these util functions private to distinguish with the collective APIs we defined?